server/src/service-layer/controllers/PaymentController.ts

```diff
@@ -176,4 +184,182 @@ export class PaymentController extends Controller {
+        }
+      }
+    }
+
+    const recentActiveSessions = await stripeService.getRecentActiveSessions(
+      CheckoutTypeValues.BOOKING,
+      30
+    )
+
+    const bookingSlotService = new BookingSlotService()
+    const bookingDataService = new BookingDataService()
+
+    const bookingSlots =
+      await bookingSlotService.getBookingSlotsBetweenDateRange(
+        startDate,
+        endDate
+      )
+    // Create a date range list to validate against
+    const daysList = datesToDateRange(
+      firestoreTimestampToDate(startDate),
+      firestoreTimestampToDate(endDate)
+    )
+
+    if (bookingSlots.length !== daysList.length) {
+      this.setStatus(401)
+      return {
+        error: "No booking slot available for one or more dates."
+      }
+    }
+    // iterate through the active stripe sessions and check metadata, reduce the bookingSlots available
+    for (const session of recentActiveSessions) {
+      const bookingSlotMetadata = session.metadata.booking_slots
+      if (bookingSlotMetadata) {
+        for (const bookingSlotId of bookingSlotMetadata) {
+          const bookingSlot = bookingSlots.find(
+            (slot) => slot.id === bookingSlotId
+          )
+          if (bookingSlot) {
+            bookingSlot.max_bookings -= 1
+          }
+        }
+      }
+    }
+    try {
+      Promise.all(
+        daysList.map(async (dateToValidate: Date, index: number) => {
+          // booking slot id and max booking slots
+          const { id, max_bookings } = bookingSlots[index]
+          if (max_bookings < 1) {
+            this.setStatus(401)
+            return {
+              error: "No booking slot available for one or more dates."
+            }
+          }
+          const bookings = await bookingDataService.getBookingsBySlotId(id)
+          if (bookings.some((booking) => booking.user_id === uid)) {
+            this.setStatus(409)
+            return {
+              error: "User already has a booking for this date"
+            }
+          }
+          return dateToValidate
+        })
+      )
+    } catch (e) {
+      this.setStatus(501)
+      return {
+        error: "Something went wrong"
+      }
+    }
+    // get requiredBookingType
+    let requiredBookingType: LodgeBookingTypeValues
+    if (numberOfDays === 1 && [6, 7].includes(daysList[0].getUTCDay())) {
+      requiredBookingType = LodgeBookingTypeValues.Weekend
+    } else {
+      requiredBookingType = LodgeBookingTypeValues.Weekday
+    }
+
+    // implement pricing logic
+    const requiredBookingProducts = await stripeService.getProductByMetadata(
+      LODGE_BOOKING_TYPE_KEY,
+      requiredBookingType
+    )
+    const requiredBookingProduct = requiredBookingProducts.find(
+      (product) => product.active
+    )
+    const { default_price } = requiredBookingProduct
+    const clientSecret = await stripeService.createCheckoutSession(
+      uid,
+      `${process.env.FRONTEND_URL}/return?session_id={CHECKOUT_SESSION_ID}`,
```

Comment:
we should also add date strings as url params so the success page can display the information to the user